File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(socket: socket.socket, size):
    data = b""
    while len(data) < size:
        bytes_read = socket.recv(size - len(data))
        if not bytes_read:
            raise ConnectionError("Socket connection closed before receiving all data")
        data += bytes_read
    return data


def send_all(socket: socket.socket, bytes):
    total_bytes = len(bytes)

    while total_bytes > 0:
        bytes_write = socket.send(bytes)
        total_bytes -= bytes_write
        bytes = bytes[bytes_write:]

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"io"
)

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	totalBytes := len(bytes)

	for totalBytes > 0 {
		bytesWrite, err := socket.Write(bytes)
		if err != nil {
			return err
		}
		totalBytes -= bytesWrite
		bytes = bytes[bytesWrite:]
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	data := make([]byte, 0, size)
	buff := make([]byte, size)
	for len(data) < size {
		bytesRead, err := socket.Read(buff[:size-len(data)])
		if err != nil {
			return nil, err
		}
		data = append(data, buff[:bytesRead]...)
	}
	return data, nil
}
